### PR TITLE
add support for DCP and FSDP async save

### PR DESCRIPTION
**Pull request by dimapihtar**
# What does this PR do ?
Implements async save strategy for `fsdp_dtensor` & `torch_dcp` checkpointing formats using NVRx async strategy.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/training/checkpointing.py [hunk 1]
```diff
@@ -65,6 +65,22 @@
...
+try:
+    from nvidia_resiliency_ext.checkpointing.async_ckpt.core import AsyncCallsQueue, AsyncRequest
+    from nvidia_resiliency_ext.checkpointing.async_ckpt.filesystem_async import FileSystemWriterAsync
+    from nvidia_resiliency_ext.checkpointing.async_ckpt.state_dict_saver import (
+        init_checkpoint_metadata_cache,
+        save_state_dict_async_finalize,
+        save_state_dict_async_plan,
+    )
+    HAVE_NVRX = True
+    async_queue = AsyncCallsQueue()
```
Comment (sbak5) [inline megatron/training/checkpointing.py]: Use `training/async_utils.py`

File: megatron/training/checkpointing.py [hunk 2]
```diff
@@ -668,11 +681,31 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
...
+            if args.async_save:
+                planner = torch.distributed.checkpoint.DefaultSavePlanner()
+                coordinator_rank = 0
+                use_msc = True if MultiStorageClientFeature.is_enabled() else False
+                fs_storage_writer = FileSystemWriterAsync(
+                    checkpoint_name, thread_count=args.dist_ckpt_workers, use_msc=use_msc
+                )
+
+                save_state_dict_ret = save_state_dict_async_plan(
+                    state_dict, fs_storage_writer, None, coordinator_rank, planner=planner, enable_cache=True
```
Comment (sbak5) [inline megatron/training/checkpointing.py]: we enable cache only when ckpt_assume_constant_structure is set for non-FSDP path.

Comment (dimapihtar) [reply megatron/training/checkpointing.py]: > we enable cache only when ckpt_assume_constant_structure is set for non-FSDP path.

so it should be False in this case?

Comment (sbak5) [reply megatron/training/checkpointing.py]: I think you need to pass `args.ckpt_assume_constant_structure` to enable_cache.

File: megatron/training/checkpointing.py [hunk 3]
```diff
@@ -668,11 +681,31 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
...
+                print_rank_0(
+                    "nvidia-resiliency-ext package is not installed. Async strategy will be disabled. "
+                    "Please, install nvidia-resiliency-ext package to use async strategy."
+                )
+                args.async_save = False
+
+            if args.async_save:
+                planner = torch.distributed.checkpoint.DefaultSavePlanner()
+                coordinator_rank = 0
+                use_msc = True if MultiStorageClientFeature.is_enabled() else False
```
Comment (sbak5) [inline megatron/training/checkpointing.py]: there's an argument for this `args.enable_msc`

File: megatron/training/checkpointing.py [hunk 4]
```diff
@@ -2069,3 +2101,16 @@ def load_biencoder_checkpoint(model, only_query_model=False,
         print(' successfully loaded {}'.format(checkpoint_name))
 
     return model
+
+
+def _get_save_and_finalize_callbacks(writer, save_state_dict_ret) -> NVRxAsyncRequest:
```
Comment (sbak5) [inline megatron/training/checkpointing.py]: It's worth placing this under `training/async_utils.py`

**Review by sbak5 — APPROVED** [review]

**Comment by sbak5** [comment]
https://github.com/NVIDIA/Megatron-LM/actions/runs/23875869228/job/69625854743?pr=4027
This test doesn't load from the previous checkpoint (finalize callback may not be called correctly?)
It's worth running this locally to see the expected behavior happens with async_save enabled.

**Comment by dimapihtar** [comment]
> https://github.com/NVIDIA/Megatron-LM/actions/runs/23875869228/job/69625854743?pr=4027 This test doesn't load from the previous checkpoint (finalize callback may not be called correctly?) It's worth running this locally to see the expected behavior happens with async_save enabled.

I did some manual tests locally for `fsdp_dtensor` and it works fine for me. I mean I was able to load from checkpoint as well.

File: megatron/training/checkpointing.py (hunk 4 above)
Comment (dimapihtar) [reply megatron/training/checkpointing.py]: done

**Review by cspades — DISMISSED** [review]
Resumption test looks fine, less context on how NVRx plugs in but LGTM.

**Comment by sbak5** [comment]
https://github.com/NVIDIA/Megatron-LM/actions/runs/23908416313/job/69731382097?pr=4027

loading doesn't happen between runs. 

> > https://github.com/NVIDIA/Megatron-LM/actions/runs/23875869228/job/69625854743?pr=4027 This test doesn't load from the previous checkpoint (finalize callback may not be called correctly?) It's worth running this locally to see the expected behavior happens with async_save enabled.
> 
> I did some manual tests locally for `fsdp_dtensor` and it works fine for me. I mean I was able to load from checkpoint as well.



> > https://github.com/NVIDIA/Megatron-LM/actions/runs/23875869228/job/69625854743?pr=4027 This test doesn't load from the previous checkpoint (finalize callback may not be called correctly?) It's worth running this locally to see the expected behavior happens with async_save enabled.
> 
> I did some manual tests locally for `fsdp_dtensor` and it works fine for me. I mean I was able to load from checkpoint as well.

https://github.com/NVIDIA/Megatron-LM/actions/runs/23908416313/job/69731382097?pr=4027

loading doesn't happen between runs. the text file keeping the latest checkpoint iteration seems missing on this CI.

**Comment by cspades** [comment]
@sbak5 Could you also summarize for me if NVRx is a "black box" async checkpointing tool? Did we have to change any Megatron-FSDP <> MLM code related to checkpointing to make this work?

Here's some context from me - Megatron-FSDP has sharded main weights stored persistently in a collection of buffers managed by `ParamAndGradBuffer`. As long as the async checkpointing is synchronized correctly with the optimizer step and uses those sharded main weights, it should be fine, but I don't see the full context from this PR.

Tests look fine though, just slightly paranoid and want to make sure there's no issues.

**Comment by sbak5** [comment]
> @sbak5 Could you also summarize for me if NVRx is a "black box" async checkpointing tool? Did we have to change any Megatron-FSDP <> MLM code related to checkpointing to make this work?
> 
> Here's some context from me - Megatron-FSDP has sharded main weights stored persistently in a collection of buffers managed by `ParamAndGradBuffer`. As long as the async checkpointing is synchronized correctly with the optimizer step and uses those sharded main weights, it should be fine, but I don't see the full context from this PR.
> 
> Tests look fine though, just slightly paranoid and want to make sure there's no issues.

NVRx saves the same state dict in the same format but asynchronously. I mean the unit test is supposed to load checkpoints between runs, which is not happening. 

https://github.com/NVIDIA/Megatron-LM/blob/159e347203d44ef808fd084ba2982900975ab538/tests/functional_tests/test_cases/gpt/gpt3_mcore_te_tp2_zp_z3_resume_fsdp_dtensor/model_config.yaml#L53
This test seems to be set not resuming from checkpoints. @cspades can you share why this test is set to regular even though this test is saving checkpoints?

File: megatron/training/checkpointing.py [hunk 5]
```diff
@@ -668,11 +678,30 @@ def save_checkpoint(iteration, model, optimizer, opt_param_scheduler, num_floati
             if ckpt_format == "fsdp_dtensor":
                 state_dict = preprocess_fsdp_dtensor_state_dict(args, state_dict, model[0])
 
-            fs_storage_writer = torch.distributed.checkpoint.FileSystemWriter(checkpoint_name)
-            torch.distributed.checkpoint.save(
-                state_dict=state_dict,
-                storage_writer=fs_storage_writer,
-            )
+            if args.async_save and not HAVE_NVRX:
+                print_rank_0(
```
Comment (jaredcasper) [inline megatron/training/checkpointing.py]: This should be an error. If they requested async save and we can't do it then it's an error in the config/environment.

**Comment by cspades** [comment]
> This test seems to be set not resuming from checkpoints. @cspades can you share why this test is set to regular even though this test is saving checkpoints?

This was a change introduced to fix some unknown segfault issue: https://github.com/dimapihtar/Megatron-LM/commit/551b7347c619038432c1d406e9e5bfa18f1eabd4

> ci: Fix segfaults (maybe)

@ko3n1g if you can chime in, but otherwise, I can fix this in another PR if the ckpt-resume doesn't work. @dimapihtar will help do a test in this PR so I can find out early, we want checkpoint resumption to be tested.

File: megatron/training/checkpointing.py (hunk 5 above)
Comment (dimapihtar) [reply megatron/training/checkpointing.py]: fixed

**Review by jaredcasper — APPROVED** [review]

**Review by cspades — APPROVED** [review]
The ckpt resumption test is known to fail, I'll fix in a separate PR.

**Review by cuichenx — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
